shell/shell.go

```diff
@@ -57,3 +58,37 @@ func RmLink(last, path string) (string, error) {
 func AddLink(last, path, hash string) (string, error) {
 	return Exec("object", "patch", "add-link", last, path, hash)
 }
+
+func keyExists(key string) bool {
+	keys, err := Exec("key", "list", "-l")
+	if err != nil {
+		return false
+	}
+	lines := strings.Split(keys, "\n")
+	for _, line := range lines {
+		keyNames := strings.Split(line, " ")
+		for _, keyName := range keyNames {
+			if keyName != "" && keyName == key {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// Publish - Publishes the ipfs hash to the provided ipns key
+func Publish(cmd *exec.Cmd, hash, key string) (*exec.Cmd, error) {
+	if !keyExists(key) {
+		return nil, errors.New("Key does not exist")
+	}
+	if cmd != nil {
+		cmd.Process.Kill()
+	}
+
+	cmd = exec.Command("ipfs", "name", "publish", hash)
```

Comment:
what is the reason you don't use `Exec("name", "publish", hash)`?

Reply:
I need *exec.Cmd object to save in the main function. In order to track that if the old process is already running or not. Exec doesn't return *exec.Cmd object.

Reply:
please use a goroutine and follow the API, using only single process

Reply:
Okay